Code standards update May 2024. (#6684)
SkriptLang/Skript: code-conventions.md

Code contributed must be licensed under GPLv3, by **you**.
We expect that any code you contribute is either owned by you or you have explicit permission to provide and license it to us.

Licenses do not need to be printed in individual files (or packages) unless the licence applying to the code in

> that file (or package) deviates from the licence scope of its containing package.

Author: This is just the update for the standards, that can be done separately ~~because I don't know how~~.

> that file (or package) deviates from the licence scope of its containing package.

Reviewer: i can commit the changes if you want

> that file (or package) deviates from the licence scope of its containing package.

Author: You may do so :)

If, in the future, we need to re-license contributed code, we will contact all contributors involved.
If we need to remove or alter contributed code due to a licensing issue we will attempt to notify its contributor.

## Code Style

### Formatting
* Imports should be grouped together by type (e.g. all `java.lang...` imports together)
  * Following the style of existing imports in a class is encouraged, but not required
  * Wildcard `*` imports are permitted (as long as they do not interfere with existing imports), e.g. `java.lang.*`.
* Tabs, no spaces (unless in code imported from other projects)
	- No tabs/spaces in empty lines
* No trailing whitespace
* At most 120 characters per line
  	- In Javadoc/multiline comments, at most 80 characters per line
* When statements consume multiple lines, all lines but the first have two tabs of additional indentation
	- The exception to this is breaking up conditional statements (e.g. `if (x || y)`) where the
    condition starts may be aligned
* Each class begins with an empty line
* No squeezing of multiple lines of code on a single line
* Separate method declarations with empty lines

>   	- Empty line after last method in a class is *not* required

Author: I didn't change anything here it's just the indentation.

  	- Otherwise, empty line before and after method is a good rule of thumb
* If fields have Javadoc, separate them with empty lines
* Use empty lines liberally inside methods to improve readability
* Use curly brackets to start and end most blocks

>   	- When a block contains a single statement, they may be omitted

Author: The if-block wouldn't contain a single statement then, so it couldn't be omitted.

## Language Features

### Compatibility
[//]: # (To be updated after feature/2.9 for Java 17)
* Contributions should maintain Java 11 source/binary compatibility, even though compiling Skript requires Java 21
  - Users must not need JRE newer than version 11
* Versions up to and including Java 21 should work too
  - Please avoid using unsafe reflection
* It is recommended to make fields final, if they are effectively final
* Local variables and method parameters should not be declared final unless used in anonymous classes, lambdas
or try-with-resources sections where their immutability is necessary
* Methods should be declared final only where necessary
* Use `@Override` whenever applicable

>   - `@Override` annotations may be omitted to prevent compilation errors when something overrides only

Reviewer: ```suggestion
  - They may be omitted to prevent compilation errors when something overrides only
```
I don't think we need to say "override annotations" again

  on a version-dependent basis (e.g. in Library XYZ version 2 we override `getX()` but in version 3 it's
  gone, and we call it ourselves)

### null-ness
* We use **JetBrains** Annotations for specifying null-ness and method contracts.
  * If editing a file using a different annotation set (e.g. Javax, Eclipse Sisu, Bukkit)
  these should be replaced with their JetBrains equivalent.
  * The semantics for JetBrains Annotations are strict _and should be observed!_
    * Many IDEs have built-in compiler-level support for these, and can even be set to produce strict
    errors when an annotation is misused; do not misuse them.
  * **`@NotNull`**
    * > An element annotated with NotNull claims null value is forbidden to return (for methods), 
    pass to (parameters) and hold (local variables and fields).

> 	* Something is `@NotNull` iff it is never null from its inception (new X) to its garbage collection, 

Reviewer 2: ```suggestion
	* Something is `@NotNull` if it is never null from its inception (new X) to its garbage collection, 
```

Author: <img width="327" alt="image" src="https://github.com/SkriptLang/Skript/assets/14147477/00c5bf8d-ae78-485d-a673-2ed8619bd626">

Reviewer 2: Ah, I never knew about this

    i.e. there is no point in time at which the value could ever be null.
  * **`@Nullable`**
    * > An element annotated with Nullable claims null value is perfectly valid to return (for methods), 
	  > pass to (parameters) or hold in (local variables and fields). 
	  > 
	  > By convention, this annotation applied only when the value should always be checked 
	  > against null because the developer could do nothing to prevent null from happening.

> 	* Something is `@Nullable` iff there is _absolutely no way of determining_ (other than checking its 

Reviewer 2: ```suggestion
	* Something is `@Nullable` if there is _absolutely no way of determining_ (other than checking its 
```
